aten/src/ATen/BatchingRules.cpp

```diff
@@ -0,0 +1,31 @@
+#include <ATen/BatchingRules.h>
+#include <ATen/ATen.h>
+
+namespace at {
```

Comment:
I think what would be super useful is to include, at the top of this level (or some relevant file), some high level explanation about how to go about writing one of these batching rules. I bet some of this is in your design docs and just needs to be copy pasted here. Some questions an average user might ask:

* What does a batching rule function do?
* Can I reuse existing batching rule functions? When do I have to write a new one?
* Is there a recipe for writing a batching rule function?

Reply:
Put differently, we should give users some abstract model to work with. The abstract model should be obviously correct but not implementable, and then can be used to determine if the implementable version. I'm personally partial to the "whenever you do a vmap, this causes a broadcast dimension to be added to every non-vmap'ed tensor in scope" explanation.